### PR TITLE
Added evaluation with testset #93

### LINKED ISSUE #93: Developing a training pipeline to handle data training, testing, and evaluation and assessment of model performance with K-fold cross-validation
### Motivation

* [x] Data samples are 20 images and labels from [mobius dataset within the github repo](https://github.com/oocular/ready/tree/main/data/mobious/sample-frames/test3000x1700) to accelerate model development. DONE for @mxochicale by creating branch 93-adding-extra-samples
* [ ] Splitting data into train, test and evaluation to develop an appropriate training pipeline. IN PROGRESS for @stoppedtime24 
  * [x] Step 1. Adding Training and validation into the main loop as in this blog https://stanford.edu/~shervine/blog/pytorch-how-to-generate-data-parallel. 
  * [x] Step 2. Adding evaluation with testset as in https://machinelearningmastery.com/training-a-pytorch-model-with-dataloader-and-dataset/
  * [x] Step 3. Loading trained model and evaluate model as shown here https://docs.pytorch.org/tutorials/beginner/saving_loading_models.html?highlight=eval
  * [ ] Review this reference https://debuggercafe.com/saving-and-loading-the-best-model-in-pytorch/
  * Other references: 
    * See how these functions def make_train_dataset, def make_val_dataset, def make_test_dataset are defined here:   https://www.kaggle.com/code/suchith0312/pytorch-dataloader-testing
    * [`train_dataset, test_dataset = torch.utils.data.random_split(full_dataset, [0.8, 0.2])`](https://stackoverflow.com/questions/50544730/how-do-i-split-a-custom-dataset-into-training-and-test-datasets)
  * This might be a clear example 
* [ ] Using K-fold cross validation to asses model performance. See this [reference](https://discuss.pytorch.org/t/using-k-fold-cross-validation-to-train-my-model/196288/2) for `ConcatDataset` and `sklearn.model_selection import KFold`
* [ ] Train model with A100 Nvidia GPU (80GB)
* [ ] Run inference in a local GPU-model (maybe not)?  

### Additional context

* Current [train_unet_with_mobious](https://github.com/oocular/ready/blob/main/src/ready/apis/train_mobious.py) use the whole dataset of 1143 samples of the dataset using [MobiousDataset](https://github.com/oocular/ready/blob/main/src/ready/utils/datasets.py)
* Using [cross validation](https://stackoverflow.com/questions/62954433/how-to-predict-with-the-test-dataset-while-using-cross-validation) for model performance assessment is one good initial step but also you can use it for model tuning (i.e. hyperparameter search).
* Reference describing k-fold-cross-validation:
  * https://www.aionlinecourse.com/tutorial/machine-learning/k-fold-cross-validation
  * https://knowledge.dataiku.com/latest/ml-analytics/ml-concepts/concept-model-validation.html
  * https://stackoverflow.com/questions/58391295/how-is-cross-validation-performed-and-how-gridsearchcv-specifically
  * https://www.mdpi.com/2072-4292/17/7/1256
  * https://www.frontiersin.org/journals/bioengineering-and-biotechnology/articles/10.3389/fbioe.2025.1580502/full
  * Ensembling, Blending & Stacking: https://www.youtube.com/watch?v=TuIgtitqJho

## Notes.
* ...

### PR DESCRIPTION
**Pull request by stoppedtime24**
## Description
Fixes #93 : Evaluating models using testset.

## Type of change
Please delete options accordingly to the description.

- [x] New feature (non-breaking change which adds functionality)

### Suggested Checklist

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation 
- [x] My changes generate no new warnings 
- [x] I have commented my code, particularly in hard-to-understand areas 
- [ ] I have read the [`CONTRIBUTING`](https://github.com/UCL/ready/blob/main/CONTRIBUTING.md) docs
- [x] My code passes, following the style guidelines `pre-commit run -a` 
- [ ] My code is properly tested with `pytest -sv tests`
- [x] I have requested review to this PR.
- [x] I have addressed and marked as resolved all the review comments in my PR.
- [x] Make sure your branch is up-to-date with main branch. See step 4 in [`CONTRIBUTING`](https://github.com/UCL/ready/blob/main/CONTRIBUTING.md#pull-request-pr-and-merge-to-main-branch).
- [ ] Finally, I have selected `squash and merge`

### CODE REVIEW

File: src/ready/apis/train_mobious.py [hunk 1]
```diff
@@ -488,6 +491,29 @@ def main(args):
     endtime = time.time()
     elapsedtime = endtime - starttime
     logger.info(f"Elapsed time for the training loop: {elapsedtime} (sec)")
+    
+    # Evaluating model using test_set
+
+    logger.info("Commencing evaluation.")
```
[the remark this answers is not in the document]
Comment (mxochicale) [reply src/ready/apis/train_mobious.py:497]: oh, the configuration file `config_train_unet_with_mobious.yaml` should look like this:
```
model:
  models_path: "datasets/ready/mobious/models"  # Local and server paths
  checkpoint_path: "datasets/?"  # TODO: specify the checkpoint path
  model_name_for_eval: "model_name_date_etc.pth"
  model_evaluation_with_load_model_flag: true  # If true, skip training and validation, and load 'model_name_for_eval' for evaluation only.
                                               # If false, follow the standard workflow: run training and validation, then perform evaluation.
```
Then call and use those variables in your train_mobious.py.

[the remark this answers is not in the document]
Comment (mxochicale) [reply src/ready/apis/train_mobious.py:497]: Checkpoints are described more formally [here](https://docs.pytorch.org/tutorials/beginner/saving_loading_models.html#saving-loading-a-general-checkpoint-for-inference-and-or-resuming-training) and something we address and tick some of the points at https://github.com/oocular/ready/issues/80

File: src/ready/apis/train_mobious.py [hunk 2]
```diff
@@ -261,258 +261,252 @@ def main(args):
...
-            
-            # Compares how many elements in predicted and labels tensors match
-            # and counts them all 
-            num_matches += (predicted == labels.to(device)).sum().item()
 
-            # Total number of elements in label tensor
-            total_elements += torch.numel(labels)   
-     
-    test_accuracy = (num_matches / total_elements) * 100
+    test_accuracy = evaluate_model(model, test_loader, device)
```
Comment (stoppedtime24) [inline src/ready/apis/train_mobious.py:508]: Save accuracy in .csv file

File: configs/models/unet/config_train_unet_with_mobious.yaml [hunk 3]
```diff
@@ -23,6 +23,13 @@ datasets_splitting_ratios: # Ensure sum of all ratios = 1.
     validation_set : 0.15
     test_set : 0.15
 
+pretrained_model: 
+    models_path : "datasets/ready/mobious/models" # LOCAL and SERVER paths
+    checkpoint_path : "datasets/?"
+    model_name_for_eval : "weights_21-Jul-2025_15-05-30.pth" # weights_[date]_[time].pth
```
Comment (stoppedtime24) [inline configs/models/unet/config_train_unet_with_mobious.yaml:29]: Add comment to tell people model must be inputted
